File: ipv4_count/src/jmh/java/dev/aspid812/ipv4_gen/IPv4RandomGeneratorApp.java
package dev.aspid812.ipv4_gen;

import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

public class IPv4RandomGeneratorApp {

	public static final int EXIT_OK = 0;
	public static final int EXIT_FATAL = -1;

	public static Pattern POOL_SIZE_OPTION = Pattern.compile("-p(?<value>\\d+)");
	public static Pattern SAMPLE_SIZE_OPTION = Pattern.compile("-n(?<value>\\d+)(?<mult>[KMB]?)");

	public static int POOL_SIZE_DEFAULT = 64_000;
	public static long SAMPLE_SIZE_DEFAULT = Long.MAX_VALUE;    // Effectively endless

	private static long multiplier(String abbr) {
		return switch (abbr) {
			case "" -> 1L;
			case "K" -> 1_000L;
			case "M" -> 1_000_000L;
			case "B" -> 1_000_000_000L;
			default -> throw new IllegalArgumentException("Unrecognized option");
		};
	}

	private static Stream<Matcher> filterOptions(String[] options, Pattern pattern) {
		return Arrays.stream(options)
			.map(pattern::matcher)
			.filter(Matcher::matches);
	}

	public static IPv4RandomGeneratorApp forCommandLine(String... options) {
		var poolSize = filterOptions(options, POOL_SIZE_OPTION)
			.mapToInt(m -> Integer.parseInt(m.group("value")))
			.findAny()
			.orElse(POOL_SIZE_DEFAULT);
		var sampleSize = filterOptions(options, SAMPLE_SIZE_OPTION)
			.mapToLong(m -> Long.parseLong(m.group("value")) * multiplier(m.group("mult")))
			.findAny()
			.orElse(SAMPLE_SIZE_DEFAULT);
		return new IPv4RandomGeneratorApp(poolSize, sampleSize);
	}

	final int poolSize;
	final long sampleSize;

	public IPv4RandomGeneratorApp(int poolSize, long sampleSize) {
		this.poolSize = poolSize;
		this.sampleSize = sampleSize;
	}

	public int run(OutputStream output, PrintStream logger) {
		try {
			var engine = IPv4RandomGenerator.create(poolSize);
			engine.sample(sampleSize, output);
		}
		catch (IOException ex) {
			ex.printStackTrace(logger);
			return EXIT_FATAL;
		}

		return EXIT_OK;
	}

	public static void main(String[] args) {
		var application = IPv4RandomGeneratorApp.forCommandLine(args);
		var status = application.run(System.out, System.err);
		if (status != EXIT_OK) {
			System.exit(status);
		}
	}
}

File: ipv4_count/src/jmh/kotlin/dev/aspid812/ipv4_gen/IPv4RandomGeneratorApp.kt
package dev.aspid812.ipv4_gen

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.io.PrintStream

import dev.aspid812.common.Application
import dev.aspid812.common.Application.EXIT_FATAL
import dev.aspid812.common.Application.EXIT_OK
import dev.aspid812.ipv4_gen.util.Multiplier


class IPv4RandomGeneratorApp(
	val poolSize: Int,
	val sampleSize: Long
) : Application {

	companion object {
		const val POOL_SIZE_DEFAULT = IPv4RandomGenerator.RECOMMENDED_POOL_SIZE
		const val SAMPLE_SIZE_DEFAULT = Long.MAX_VALUE      // Effectively endless

		val poolSizeOption   = Regex("-p(?<value>\\d+)")
		val sampleSizeOption = Regex("-n(?<value>\\d+[KMB]?)")

		@JvmStatic
		fun forCommandLine(vararg options: String): IPv4RandomGeneratorApp {
			fun <T> Sequence<T>.singleOrNone() =
				this.reduceOrNull { _, _ -> throw IllegalArgumentException("Repeated option") }

			val poolSize = options.asSequence()
				.mapNotNull { poolSizeOption.matchEntire(it)?.destructured }
				.map { (value) -> value.toInt() }
				.singleOrNone() ?: POOL_SIZE_DEFAULT
			val sampleSize = options.asSequence()
				.mapNotNull { sampleSizeOption.matchEntire(it)?.destructured }
				.map { (value) -> Multiplier.parseLong(value) }
				.singleOrNone() ?: SAMPLE_SIZE_DEFAULT
			return IPv4RandomGeneratorApp(poolSize, sampleSize)
		}
	}

	private val engine = IPv4RandomGenerator(poolSize)

	override fun run(input: InputStream?, output: PrintStream?, logger: PrintStream?) =
		run(
			output = requireNotNull(output),
			logger = requireNotNull(logger)
		)

	fun run(output: OutputStream, logger: PrintStream): Int {
		try {
			engine.sample(sampleSize).draw(output)
		}
		catch (ex: IOException) {
			ex.printStackTrace(logger)
			return EXIT_FATAL
		}

		return EXIT_OK
	}
}

fun IPv4RandomGeneratorApp(vararg options: String): IPv4RandomGeneratorApp {
	//TODO: Catch any exception caused by invalid options and wrap it to IAE
	return IPv4RandomGeneratorApp.forCommandLine(*options)
}


fun main(args: Array<String>) {
	val application = IPv4RandomGeneratorApp.forCommandLine(*args)
	val status = application.run(System.out, System.err)
	if (status != EXIT_OK) {
		kotlin.system.exitProcess(status)
	}
}
